rust/chains/abacus-ethereum/src/multisig_ism.rs

```diff
@@ -0,0 +1,237 @@
+#![allow(clippy::enum_variant_names)]
+#![allow(missing_docs)]
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::{Duration, SystemTime};
+
+use abacus_core::accumulator::merkle::Proof;
+use async_trait::async_trait;
+use ethers::abi::Token;
+use ethers::providers::Middleware;
+use ethers::types::{Selector, H160, H256, U256};
+use eyre::Result;
+use std::collections::hash_map::Entry::Occupied;
+use std::hash::Hash;
+use tokio::sync::Mutex;
+
+use abacus_core::{
+    AbacusAbi, AbacusContract, ChainCommunicationError, ContractLocator, MultisigIsm,
+    MultisigSignedCheckpoint,
+};
+
+use crate::contracts::multisig_ism::{MultisigIsm as EthereumMultisigIsmInternal, MULTISIGISM_ABI};
+use crate::trait_builder::MakeableWithProvider;
+
+#[derive(Debug)]
+struct Timestamped<Value> {
+    t: SystemTime,
+    value: Value,
+}
+
+impl<Value> Timestamped<Value> {
+    fn new(value: Value) -> Timestamped<Value> {
+        Timestamped {
+            t: SystemTime::now(),
+            value,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ExpiringCache<Key, Value>
+where
+    Key: Eq + Hash,
+{
+    expiry: Duration,                        // cache endurance
+    cache: HashMap<Key, Timestamped<Value>>, // hashmap containing references to cached items
+}
+
+impl<Key, Value> ExpiringCache<Key, Value>
+where
+    Key: Copy + Eq + Hash + tracing::Value,
+    Value: Clone,
+{
+    pub fn new(expiry: Duration) -> ExpiringCache<Key, Value> {
+        ExpiringCache {
+            expiry,
+            cache: HashMap::new(),
+        }
+    }
+
+    pub fn put(&mut self, key: Key, value: Value) {
+        self.cache.insert(key, Timestamped::new(value));
+    }
+
+    pub fn get(&mut self, key: Key) -> Option<&Value> {
```

Comment:
does this need to be mut?

Reply:
It did when I was using `self.cache.entry` since entry returns a mut, but `get` doesn't, so no